File: DSA/Stack/Code/PostfixToPrefix.py
# Postfix to Prefix in Python

def is_operator(c):
    return c in ['+', '-', '*', '/']

def postfix_to_prefix(postfix):
    stack = []
    
    for c in postfix:
        if is_operator(c):
            op2 = stack.pop()
            op1 = stack.pop()
            
            temp = c + op1 + op2
            stack.append(temp)
        else:
            stack.append(c)
    
    return stack[0]

# Test
postfix = "ab+cd*-"

print("=== Postfix to Prefix ===\n")
print(f"Postfix Expression: {postfix}")

prefix = postfix_to_prefix(postfix)

print(f"Prefix Expression: {prefix}\n")
print("Complexity: O(n)")

File: DSA/Stack/Code/PostfixToPrefix.java
// Postfix to Prefix in Java

import java.util.Stack;

public class PostfixToPrefix {
    static boolean isOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/');
    }
    
    static String postfixToPrefix(String postfix) {
        Stack<String> stack = new Stack<>();
        
        for (char c : postfix.toCharArray()) {
            if (isOperator(c)) {
                String op2 = stack.pop();
                String op1 = stack.pop();
                
                String temp = c + op1 + op2;
                stack.push(temp);
            } else {
                stack.push(String.valueOf(c));
            }
        }
        
        return stack.pop();
    }
    
    public static void main(String[] args) {
        String postfix = "ab+cd*-";
        
        System.out.println("=== Postfix to Prefix ===\n");
        System.out.println("Postfix Expression: " + postfix);
        
        String prefix = postfixToPrefix(postfix);
        
        System.out.println("Prefix Expression: " + prefix + "\n");
        System.out.println("Complexity: O(n)");
    }
}
